File: testdata/fixtures/elixir/CWE-502.ex
# CWE-502: Deserialization via binary_to_term
defmodule Fixture.CWE502 do
  def vuln1(data) do
    # ruleid: elixir-deserialization
    :erlang.binary_to_term(data)
  end

  def vuln2(conn) do
    payload = conn.params["data"]
    # ruleid: elixir-deserialization
    :erlang.binary_to_term(payload)
  end

  def vuln3(user_input) do
    # ruleid: elixir-deserialization
    :erlang.binary_to_term(user_input, [:safe])
  end

  def safe do
    # ok: elixir-deserialization
    :erlang.binary_to_term("fixed_binary")
  end

  def safe2 do
    # ok: elixir-deserialization
    :erlang.binary_to_term("fixed")
  end
end

File: testdata/fixtures/go/CWE-502.go
package fixture

// CWE-502: Deserialization of Untrusted Data via gob
import (
	"bytes"
	"encoding/gob"
	"net/http"
)

func vuln1(r *http.Request) {
	// ruleid: go-deserialization
	gob.NewDecoder(r.Body).Decode(&struct{}{})
}

func vuln2(r *http.Request) {
	// ruleid: go-deserialization
	gob.NewDecoder(r.Body).Decode(&struct{}{})
}

func safe1() {
	// ok: go-deserialization
	gob.NewDecoder(bytes.NewBufferString("fixed")).Decode(&struct{}{})
}

func safe2() {
	// ok: go-deserialization
	gob.NewDecoder(bytes.NewBufferString("hello")).Decode(&struct{}{})
}
func vuln3(r *http.Request) {
    // ruleid: go-deserialization
    gob.NewDecoder(r.Body).Decode(&struct{}{})
}

func vuln4(r *http.Request) {
    q := r.URL.Query().Get("data")
    // ruleid: go-deserialization
    gob.NewDecoder(bytes.NewBufferString(q)).Decode(&struct{}{})
}

func safe3() {
    // ok: go-deserialization
    gob.NewDecoder(bytes.NewReader([]byte("fixed"))).Decode(&struct{}{})
}
